walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,108 @@
+        for(TimeRange tr : booked_times){
+            TimeRange free = TimeRange.fromStartEnd(firstTime, tr.start(), false);
+
+            // Ensures a meeting ending before the endpoint of a previous meeting cannot allow
+            // an invalid meeting to be scheduled
+            if(tr.start() + tr.duration() > firstTime)
+                firstTime = tr.start() + tr.duration();
+
+            if(free.duration() >= requestDuration)
+                possible_times.add(free);
+        }
+        TimeRange endOfDay = TimeRange.fromStartEnd(firstTime, 1440, false);
+        if(endOfDay.duration() >= requestDuration){
+            possible_times.add(endOfDay);
+        }
+        
+        return possible_times;
+    }
+
+    // This method doesn't seem efficient. I'd appreciate suggestions
+    // TODO: Refactor for efficiency
+    private ArrayList<TimeRange> checkTimes(Collection<Event> events, 
+        MeetingRequest request, boolean optionalCheck){
+
+        ArrayList<TimeRange> booked_times = new ArrayList<TimeRange>();
+        for(Event e : events){
+            //Adds the time slot to booked times if it pertains to request people
+            Set<String> people = e.getAttendees();
+            for(String person : people){
+                if(request.getAttendees().contains(person) || 
+                    (request.getOptionalAttendees().contains(person) && optionalCheck)){
+                    
+                    boolean overlap = false;
+                    for(int ii = 0; ii < booked_times.size(); ii++){
+                        TimeRange bt = booked_times.get(ii);
+                        if(bt.overlaps(e.getWhen()) || bt.end() == e.getWhen().start() || 
+                            bt.start() == e.getWhen().end()){
+                            overlap = true;
+                            TimeRange newTime = 
+                                TimeRange.fromStartEnd(Math.min(e.getWhen().start(), bt.start()),
+                                Math.max(e.getWhen().end(), bt.end()), false);
+                            booked_times.set(ii, newTime);
+                        }
+                    }
+                    if(!overlap){
+                        TimeRange newTime = e.getWhen();
+                        TimeRange temp;
+                        // Sorts meeting by start time
+                        // Can't have overlap
+                        for(int ii = 0; ii < booked_times.size(); ii++){
+                            TimeRange bt = booked_times.get(ii);
+                            if(newTime.start() < bt.start()){
+                                temp = bt;
+                                booked_times.set(ii, newTime);
+                                newTime = temp;
+                            }
+                        }
+                        booked_times.add(newTime);
+                    }
+                    break;
+                }
+            }
+        }
+        return booked_times;
```

Comment:
`O(booked_times) == O(events)` since we loop through `events (e)` and `people (p)` and then `booked_times` this ends up being `O(p*e^2)` Which i think can be improved on.

What if you collected all the relevent events first without the loop over booked_times.
`O(p*e)` to get `booked_times`.

Then, because later methods require you to have sorted non-overlapping intervals, you can sort the list of `booked_times` in `O(e*log(e))`.
Then merging overlapping intervals becomes a lot easier and can be done in 
`O(e)`
So you end up with `O(p*e + e*log(e) + e) == O(p*e + e*log(e)) < O(p*e^2)`